### PR TITLE
Feature/four channel

### PR DESCRIPTION
**Pull request by eatmore18**
Add the four-channel feature to dev. Now the model is able to train on images that have more than 3 channels, like temporal channels or motion channels.

### CODE REVIEW

File: utils/general.py [hunk 1]
```diff
@@ -960,9 +960,9 @@ def non_max_suppression(
         output[xi] = x[i]
         if mps:
             output[xi] = output[xi].to(device)
-        if (time.time() - t) > time_limit:
-            LOGGER.warning(f'WARNING ⚠️ NMS time limit {time_limit:.3f}s exceeded')
-            break  # time limit exceeded
+        # if (time.time() - t) > time_limit:
+        #     LOGGER.warning(f'WARNING ⚠️ NMS time limit {time_limit:.3f}s exceeded')
+        #     break  # time limit exceeded
```
Comment (eatmore18) [inline utils/general.py:965]: comment

File: utils/general.py [hunk 2]
```diff
@@ -960,9 +960,9 @@ def non_max_suppression(
         output[xi] = x[i]
         if mps:
             output[xi] = output[xi].to(device)
-        if (time.time() - t) > time_limit:
-            LOGGER.warning(f'WARNING ⚠️ NMS time limit {time_limit:.3f}s exceeded')
-            break  # time limit exceeded
+        # if (time.time() - t) > time_limit:
```
Comment (amnikoo) [inline utils/general.py]: We can increase time_limit instead of commenting this assertion. It's more clear.

File: utils/augmentations.py [hunk 3]
```diff
@@ -77,7 +77,7 @@ def augment_hsv(im, hgain=0.5, sgain=0.5, vgain=0.5):
         lut_val = np.clip(x * r[2], 0, 255).astype(dtype)
 
         im_hsv = cv2.merge((cv2.LUT(hue, lut_hue), cv2.LUT(sat, lut_sat), cv2.LUT(val, lut_val)))
-        cv2.cvtColor(im_hsv, cv2.COLOR_HSV2BGR, dst=im)  # no return needed
+        cv2.cvtColor(im_hsv, cv2.COLOR_HSV2BGR, dst=im.copy())  # no return needed
```
Comment (amnikoo) [inline utils/augmentations.py:80]: Do we need copy image in this line?

File: train.py [hunk 4]
```diff
@@ -112,6 +112,8 @@ def train(hyp, opt, device, callbacks):  # hyp is path/to/hyp.yaml or hyp dictio
         data_dict = data_dict or check_dataset(data)  # check if None
     train_path, val_path = data_dict['train'], data_dict['val']
     nc = 1 if single_cls else int(data_dict['nc'])  # number of classes
+    channels = data_dict['channels'] if 'channels' in data_dict else []  # the channels of data more than RGB
```
Comment (amnikoo) [inline train.py:115]: What is difference between channels and ch arguments?
It seems ch is for number of channels; but, what about channels? and where is it used?

File: utils/dataloaders.py [hunk 5]
```diff
@@ -714,13 +730,20 @@ def __getitem__(self, index):
...
             labels_out[:, 1:] = torch.from_numpy(labels)
 
         # Convert
-        img = img.transpose((2, 0, 1))[::-1]  # HWC to CHW, BGR to RGB
+        rgb = img[:,:,:3]
+        ch = img[:,:,3:]
+        rgb = rgb.transpose((2, 0, 1))[::-1]
+        ch = ch.transpose((2, 0, 1))
+        img = np.concatenate((rgb, ch),axis = 0)  # HWC to CHW, BGR to RGB
+
```
Comment (amnikoo) [inline utils/dataloaders.py:746]: These 5 line of codes repeated 3 times. It could be mudulated as a function.

File: utils/dataloaders.py [hunk 6]
```diff
@@ -734,6 +757,14 @@ def load_image(self, i):
             else:  # read image
                 im = cv2.imread(f)  # BGR
                 assert im is not None, f'Image Not Found {f}'
+                for ch_dir in self.channels:
+                    ch_path = f.replace(f.split('/')[-2],ch_dir)
+                    channel = cv2.imread(ch_path, cv2.IMREAD_UNCHANGED)
+                    if channel.ndim < 3:
```
Comment (amnikoo) [inline utils/dataloaders.py:763]: Why this condition may happen?

File: val.py [hunk 7]
```diff
@@ -154,6 +154,9 @@ def run(
 
         # Data
         data = check_dataset(data)  # check
+        
+    channels = data['channels'] if 'channels' in data else []
```
Comment (amnikoo) [inline val.py:158]: What is difference between channels and ch arguments?
It seems ch is for number of channels; but, what about channels? and where is it used?

File: utils/loss.py [hunk 8]
```diff
@@ -92,19 +95,22 @@ class ComputeLoss:
...
-    def __init__(self, model, autobalance=False):
+    def __init__(self, model, autobalance=False, uncertain=False):
+        self.uncertain = uncertain
         device = next(model.parameters()).device  # get model device
         h = model.hyp  # hyperparameters
 
         # Define criteria
         BCEcls = nn.BCEWithLogitsLoss(pos_weight=torch.tensor([h['cls_pw']], device=device))
-        BCEobj = nn.BCEWithLogitsLoss(pos_weight=torch.tensor([h['obj_pw']], device=device))
+        # BCEobj = nn.BCEWithLogitsLoss(pos_weight=torch.tensor([h['obj_pw']], device=device))
```
Comment (amnikoo) [inline utils/loss.py:105]: This commented code can be deleted.

File: utils/loss.py [hunk 9]
```diff
@@ -159,8 +168,18 @@ def __call__(self, p, targets):  # predictions, targets
                 # Append targets to text file
                 # with open('targets.txt', 'a') as file:
                 #     [file.write('%11.5g ' * 4 % tuple(x) + '\n') for x in torch.cat((txy[i], twh[i]), 1)]
-
-            obji = self.BCEobj(pi[..., 4], tobj)
+                
+            if self.g>0:
```
Comment (amnikoo) [inline utils/loss.py:172]: Why is the value of g (fl_gamma) important here and should it be taken into account?

**Review by amnikoo — COMMENTED** [review]
Ty for these codes. There were a few questions that are good to be answered and some commented codes should be deleted to make the code cleaner.
